controllers/catalogsource_controller.go

```diff
@@ -0,0 +1,182 @@
+	defaultRegistryGRPCConnectionTimeout = 10 * time.Second
+
+	eventTypeNormal  = "Normal"
+	eventTypeWarning = "Warning"
+
+	eventReasonCacheUpdated      = "BundleCacheUpdated"
+	eventReasonCacheUpdateFailed = "BundleCacheUpdateFailed"
+)
+
+type CatalogSourceReconcilerOption func(reconciler *CatalogSourceReconciler)
+
+func WithRegistryClient(registry catalogsource.RegistryClient) CatalogSourceReconcilerOption {
+	return func(reconciler *CatalogSourceReconciler) {
+		reconciler.registry = registry
+	}
+}
+
+// applyDefaults applies default values to empty CatalogSourceReconciler fields _after_ options have been applied
+func applyDefaults() CatalogSourceReconcilerOption {
+	return func(reconciler *CatalogSourceReconciler) {
+		if reconciler.registry == nil {
+			reconciler.registry = catalogsource.NewRegistryGRPCClient(defaultRegistryGRPCConnectionTimeout)
+		}
+	}
+}
+
+type CatalogSourceReconciler struct {
+	sync.RWMutex
+	client.Client
+	scheme   *runtime.Scheme
+	registry catalogsource.RegistryClient
+	recorder record.EventRecorder
+	cache    map[string]map[deppy.Identifier]*input.Entity
+}
+
+func NewCatalogSourceReconciler(client client.Client, scheme *runtime.Scheme, recorder record.EventRecorder, options ...CatalogSourceReconcilerOption) *CatalogSourceReconciler {
+	reconciler := &CatalogSourceReconciler{
+		RWMutex:  sync.RWMutex{},
+		Client:   client,
+		scheme:   scheme,
+		recorder: recorder,
+		cache:    map[string]map[deppy.Identifier]*input.Entity{},
+	}
+	// apply options
+	options = append(options, applyDefaults())
+	for _, option := range options {
+		option(reconciler)
+	}
+
+	return reconciler
+}
+
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=catalogsources,verbs=get;list;watch
+
+func (r *CatalogSourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithName("catalogsource-controller")
+	l.V(1).Info("starting")
+	defer l.V(1).Info("ending")
+
+	var catalogSource = &v1alpha1.CatalogSource{}
+	if err := r.Client.Get(ctx, req.NamespacedName, catalogSource); err != nil {
+		if errors.IsNotFound(err) {
+			r.dropSource(req.String())
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	entities, err := r.registry.ListEntities(ctx, catalogSource)
+	if err != nil {
+		r.recorder.Event(catalogSource, eventTypeWarning, eventReasonCacheUpdateFailed, fmt.Sprintf("Failed to update bundle cache from %s/%s: %v", catalogSource.GetNamespace(), catalogSource.GetName(), err))
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.updateCache(req.String(), entities)
+	r.recorder.Event(catalogSource, eventTypeNormal, eventReasonCacheUpdated, fmt.Sprintf("Successfully updated bundle cache from %s/%s", catalogSource.GetNamespace(), catalogSource.GetName()))
+
+	if isManagedCatalogSource(*catalogSource) {
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{RequeueAfter: defaultCatalogSourceSyncInterval}, nil
```

Comment:
Is `RequeueAfter` necessary? 

Does the OLMv0 catalog-operator update the catsrc status in some way when a catsrc is pivoted to a new image? If so, we can just rely on events coming in from our watch, right?

Reply:
I see now. IMO I think we can assume that if the OLMv0 catalog source API is present, so is the OLMv0 catalog operator.

Reply:
I'm assuming this is to allow an easier transition away from OLMv0 - to allow removing the v0 components beofre we've completed the v1 catalogs. It may he helpful when we're planning migration in the future.

Reply:
Migration is a good point. But I'm wondering if we'll still need at least the OLMv0 catalog operator (or equivalent)? If we don't have it that'll mean:
1. Catalog sources that define images will never end up with a catalog pod or address to connect to (its the catalog operator that manages that).
2. No catalog sources will have their status block updated.

It seems like we'll either need the OLMv0 catalog operator, or we'll need something that can migrate existing OLMv0 catalog sources to the new OLMv1 cluster catalog source.

Reply:
I'd also say the the underlying workloads of v0 CatalogSources (eg registry pods) are not(ideally, hopefully) the underlying workloads for v1 CatalogSources. 

I'm also curious if we need to make this adapter aware of different types of CatalogSources. If all of this is already fully fleshed out then sure let's go for it, but I'm not sure if it's worth solving for every type of CatalogSource, if that'll make the implementation easy. Eg in this case, if we just solve for grpc CatalogSources, [LatestImageRegistryPoll](https://github.com/operator-framework/api/blob/master/pkg/operators/v1alpha1/catalogsource_types.go#L206) is the thing that lets you know if there's been a new polling update rollout.